File: leetcode/Python/0035-e-search-insert-position.py
"""
Time: O(logn)
Space: O(1)
"""
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        start, end = 0 , len(nums)-1
        while start <= end: 
            mid = (start + end) // 2 
            if nums[mid] >= target: 
                end = mid - 1 
            elif nums[mid] < target: 
                start = mid+1 
        return start

File: leetcode/C++/0035-e-search-insert-position.cpp
class Solution {
public: 
    int searchInsert(vector<int>& nums, int target) {
        return distance(nums.begin(), lower_bound(nums.begin(), nums.end(), target));
    }
private:
    template<typename ForwardIterator, typename T>
    ForwardIterator lower_bound(ForwardIterator first, ForwardIterator last, T value) {
        while (first != last) {
            auto mid = next(first, distance(first, last) / 2);

            if (value > *mid)
                first = ++mid;
            else
                last = mid;
        }
        return first;
    }
};